### PR TITLE
Use embed for urbandictionary

### PR DESCRIPTION
**Pull request by Miroito**

### CODE REVIEW

File: src/modules/commands/discord/urbandictionary.ts [hunk 1]
```diff
@@ -61,16 +67,114 @@ export default new Command<discord.Message>({
...
+
+    setTimeout(() => {
+      definitionEmbed
+        .setFooter('Interaction ended')
+        .setColor('#1D2439')
+        .setTimestamp()
+      void embedMessage.edit({ embed: definitionEmbed })
+    }, timeout)
+
+    collector.on('collect', (reaction) => {
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:118]: There is quite a lot of duplicate code from this line below.

File: src/modules/commands/discord/urbandictionary.ts [hunk 2]
```diff
@@ -61,16 +67,114 @@ export default new Command<discord.Message>({
...
+      if (reaction.emoji.name === constants.discord.emojis.arrow_right) {
+        try {
+          void reaction.users.remove(raw.author)
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:121]: You need to use await instead of void, otherwise the error will not be caught by try/catch.

File: src/modules/commands/discord/urbandictionary.ts [hunk 3]
```diff
@@ -61,16 +67,114 @@ export default new Command<discord.Message>({
...
+        } catch (error) {
+          console.log(error)
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:123]: Just logging the error isn't enough to immediately see where it came from. Consider something like `console.error('Error removing reaction:', error)`

File: src/modules/commands/discord/urbandictionary.ts (hunk 2 above)
Comment (vostrnad) [reply src/modules/commands/discord/urbandictionary.ts:121]: Also, you probably want to remove the reaction after you've done everything else.

File: src/modules/commands/discord/urbandictionary.ts [hunk 4]
```diff
@@ -61,16 +67,114 @@ export default new Command<discord.Message>({
...
+      .setAuthor('Urban dictionary', '', 'https://www.urbandictionary.com')
+      .setThumbnail('https://i.imgur.com/A6nvY85.png')
+      .addFields(
+        {
+          name: 'Definition',
+          value: list[defN].definition.replace(re, '*'),
+        },
+        { name: 'Example', value: list[defN].example.replace(re, '*') },
+      )
+      .addField('Written on', list[defN].writtenOn.substring(0, 11), true)
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:88]: The "Written on" field looks pretty ugly if you just take the raw timestamp. You can pass that timestamp to `new Date(timestamp)` and just extract the date (we don't need time).

File: src/global/constants.ts [hunk 5]
```diff
@@ -16,6 +16,12 @@ export const constants = {
     },
     userIds: {
       alfav: '259419275457069056',
+      Miroito: '240393399386112000',
```
Comment (vostrnad) [inline src/global/constants.ts:19]: Why is this here?

File: src/global/constants.ts [hunk 6]
```diff
@@ -16,6 +16,12 @@ export const constants = {
...
+    },
+    emojis: {
+      arrow_left: '\u2b05',
+      arrow_right: '\u27a1',
+      white_check_mark: '\u2705',
```
Comment (vostrnad) [inline src/global/constants.ts:24]: Field names should be in camelCase.

File: src/modules/commands/discord/urbandictionary.ts [hunk 7]
```diff
@@ -37,7 +40,10 @@ export default new Command<discord.Message>({
       .json()
       .then((data) => {
         if (!isRecord(data)) {
-          return Promise.reject(new Error(`Unexpected query return type`))
+          return Promise.reject(
+            new Error(`Unexp    definitionEmbed.setFooter('Interaction ended').setColor('#1D2439')
+          await embedMessage.edit({ embed: definitionEmbed })ected query return type`),
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:45]: What is this?

File: src/modules/commands/discord/urbandictionary.ts [hunk 8]
```diff
@@ -51,7 +57,7 @@ export default new Command<discord.Message>({
         return camelcaseKeys(definitions, { deep: true })
       })) as Definition[]
 
-    const re = /\[(\w+(\s\w+)*?)\]/g
+    const re = /\[|\]/g
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:60]: You don't need a special variable for this.

File: src/modules/commands/discord/urbandictionary.ts [hunk 9]
```diff
@@ -61,16 +67,114 @@ export default new Command<discord.Message>({
...
+        [
+          constants.discord.emojis.arrow_left,
+          constants.discord.emojis.arrow_right,
+        ].includes(reaction.emoji.name) && user.id === raw.author.id,
+      {
+        time: timeout,
+      },
+    )
+
+    setTimeout(() => {
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:110]: Instead of using setTimeout, listen for the event `collector` fires when it stops listening for reactions.

File: src/modules/commands/discord/urbandictionary.ts [hunk 10]
```diff
@@ -61,16 +67,114 @@ export default new Command<discord.Message>({
...
+      if (reaction.emoji.name === constants.discord.emojis.arrow_right) {
+        try {
+          void reaction.users.remove(raw.author)
+        } catch (error) {
+          console.log(error)
+        }
+        defN += 1
+        if (defN + 1 > list.length) {
+          defN = 0
+        }
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:128]: See if you can simplify the code for updating defN using modular arithmetic.

**Review by vostrnad — CHANGES_REQUESTED** [review]

File: src/global/constants.ts (hunk 6 above)
Comment (Miroito) [reply src/global/constants.ts:24]: The names on discord use this convention, should I still change it to camelCase and then we eventually use a library to convert it back to their original names?

File: src/modules/commands/discord/urbandictionary.ts (hunk 7 above)
Comment (Miroito) [reply src/modules/commands/discord/urbandictionary.ts:45]: Good question maybe a paste gone wrong that I didn't notice

File: src/modules/commands/discord/urbandictionary.ts (hunk 4 above)
Comment (Miroito) [reply src/modules/commands/discord/urbandictionary.ts:88]: The substring will extract the date like "2021-08-11" isn't that good?

File: src/modules/commands/discord/urbandictionary.ts (hunk 2 above)
Comment (Miroito) [reply src/modules/commands/discord/urbandictionary.ts:121]: If I use await I need to set the function to be async, but then I have return issues.

When I type "return void" I get a "Parsing error: Expression needed" which I looked up on the internet and people don't have an answer other than update yarn usually

File: src/modules/commands/discord/urbandictionary.ts (hunk 4 above)
Comment (vostrnad) [reply src/modules/commands/discord/urbandictionary.ts:88]: I see, but you don't extract the substring when updating the embed below.

File: src/global/constants.ts (hunk 6 above)
Comment (vostrnad) [reply src/global/constants.ts:24]: Yes, you never pass these names to discord.js so use camelCase for our code.

**Also, the strings should be `'\u2b05\ufe0f'` and `'\u27a1\ufe0f'`.**

File: src/modules/commands/discord/urbandictionary.ts (hunk 2 above)
Comment (vostrnad) [reply src/modules/commands/discord/urbandictionary.ts:121]: You can have async if you create an async function within this callback and call it. No return issues.

File: src/modules/commands/discord/urbandictionary.ts (hunk 4 above)
Comment (vostrnad) [reply src/modules/commands/discord/urbandictionary.ts:88]: And it needs to be (0, 10), what you're extracting now is "2021-08-11T".

File: src/modules/commands/discord/urbandictionary.ts (hunk 8 above)
Comment (Miroito) [reply src/modules/commands/discord/urbandictionary.ts:60]: I use it 4 times and I don't want to change it 4 times if I need to make a change there

File: src/global/constants.ts (hunk 6 above)
Comment (vostrnad) [reply src/global/constants.ts:23]: Names still not camelCase.

File: src/modules/commands/discord/urbandictionary.ts (hunk 4 above)
Comment (vostrnad) [reply src/modules/commands/discord/urbandictionary.ts]: Still (0, 11).

File: src/modules/commands/discord/urbandictionary.ts [hunk 11]
```diff
@@ -72,27 +119,22 @@ export default new Command<discord.Message>({
...
-      .addField('Thumbs up', list[defN].thumbsUp, true)
-      .addField('Thumbs down', list[defN].thumbsDown, true)
-      .setFooter('Interactive')
-      .setTimestamp()
+
+    let definitionEmbed = genDefinitionEmbed(list[defN], list.length)
 
     const embedMessage = await raw.channel.send({ embed: definitionEmbed })
 
+    if (list.length === 1) {
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:127]: This check should be done before you send the embed. Sending it and immediately editing it is unnecessary.

File: src/modules/commands/discord/urbandictionary.ts [hunk 12]
```diff
@@ -28,9 +28,47 @@ export default new Command<discord.Message>({
...
+        .addField('Written on', definition.writtenOn.substring(0, 10), true)
+        .addField('Thumbs up', definition.thumbsUp, true)
+        .addField('Thumbs down', definition.thumbsDown, true)
+        .setFooter('Interactive')
+        .setTimestamp()
+
+      return definitionEmbed
+    }
+
+    function mod(n: number, m: number): number {
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:65]: Please put this in utils/math.ts and change it from `function mod ...` to `const mod = ...` if you would be so kind.

**Review by vostrnad — CHANGES_REQUESTED** [review]

File: src/modules/commands/discord/urbandictionary.ts [hunk 13]
```diff
@@ -4,6 +4,7 @@ import { isRecord } from '@app/validators/object'
 import camelcaseKeys from 'camelcase-keys'
 import discord from 'discord.js'
 import { constants } from '@app/global/constants'
+import { mod } from 'mathjs'
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:7]: Hmm, this doesn't look like `@app/utils/math` 🤔

File: src/modules/commands/discord/urbandictionary.ts [hunk 14]
```diff
@@ -61,16 +111,65 @@ export default new Command<discord.Message>({
...
+        await embedMessage.edit({ embed: definitionEmbed })
+      })()
+    })
+
+    collector.on('end', () => {
+      definitionEmbed
+        .setFooter('Interaction ended')
+        .setColor('#1D2439')
+        .setTimestamp()
+      void (async () => {
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:170]: This async wrap isn't needed. You can call `void embedMessage.edit`.

File: src/modules/commands/discord/urbandictionary.ts [hunk 15]
```diff
@@ -61,16 +111,65 @@ export default new Command<discord.Message>({
...
+
+      void (async () => {
+        await removeReaction(reaction)
+      })()
+
+      definitionEmbed = genDefinitionEmbed(list[defN], list.length)
+
+      definitionEmbed.setTimestamp()
+      void (async () => {
+        await embedMessage.edit({ embed: definitionEmbed })
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:161]: The two async wraps need to be merged into one.

**Review by vostrnad — CHANGES_REQUESTED** [review]

File: src/modules/commands/discord/urbandictionary.ts (hunk 13 above)
Comment (Miroito) [reply src/modules/commands/discord/urbandictionary.ts:7]: lol I trusted VS code and didn't look back

**Review by vostrnad — APPROVED** [review]
